### DISCUSSION TITLE
How to load/convert image as float?

### DISCUSSION THREAD

**Discussion by zanazakaryaie** [halide/Halide#4984] [General]
Hello. I want to use Halide to accelerate some convolution algorithms such as GaussianBlur and Scharr (compared to OpenCV). I found an example for 7x7 GaussianBlur [here](https://github.com/halide/CVPR2015/blob/master/blur.cpp). But it uses the old `Image<type>` to represent the image. Compiling it with the new version of Halide says:

> error: ‘Image’ was not declared in this scope

So, I changed it to `Halide::Buffer<float>`. But it seems that Halide can't load the image in float type because I get this error in runtime:

> ../include/Halide.h:9883: static void Halide::Runtime::Buffer<T, D>::assert_can_convert_from(const Halide::Runtime::Buffer<T2, D2>&) [with T2 = void; int D2 = 4; T = float; int D = 4]: Assertion `can_convert_from(other)' failed.
> Aborted (core dumped)

What should I do? I need the float version of these convolutions

**Comment by abadams** [comment]
Try using load_and_convert_image instead of load_image

**Comment by zanazakaryaie** [comment]
Thanks. I tested it and it worked faster than OpenCV. That's awesome :)

Just two more question:

1-` Tools::save_image` cannot save float output. I don't need to save. But want to see the output. Is there a way to convert `Halide::Buffer` to `cv::Mat` (OpenCV) and vice-versa without copying the data?

2- How can I see the Halide-generated C or C++ code?

**Comment by zanazakaryaie** [comment]
Also, is there any macro to set the number of threads in Halide?
OpenCV's GaussianBlur works single-threaded. I want to know that if Halide just uses one core, would it be still faster than OpenCV?

**Comment by zanazakaryaie** [comment]
I used this code to convert OpenCV's Mat to `Halide::Buffer`:

`Halide::Buffer<float> in((float *)OpenCV_Mat.data, OpenCV_Mat.rows, OpenCV_Mat.cols, OpenCV_Mat.channels());`

and used this do vice-versa:

`Mat OpenCV_Output(Halide_Output.width(), Halide_Output.height(), CV_32F, Halide_Output.data());`

But the Halide-blurred image seems strange! I didn't change the formula [here](https://github.com/halide/CVPR2015/blob/master/blur.cpp).

![Screenshot from 2020-05-31 13-00-01](https://user-images.githubusercontent.com/25720728/83348084-c5336100-a33e-11ea-8d64-45c94c906672.png)

**Comment by benzwt** [comment]
> Also, is there any macro to set the number of threads in Halide?
> OpenCV's GaussianBlur works single-threaded. I want to know that if Halide just uses one core, would it be still faster than OpenCV?

call  halide_set_num_threads(1);
before calling your function.

To my understanding, to boost up the performance, you might want to scheduler it with single-thread in mind.(without using parallel() )

**Comment by benzwt** [comment]
> I used this code to convert OpenCV's Mat to `Halide::Buffer`:
> 
> `Halide::Buffer<float> in((float *)OpenCV_Mat.data, OpenCV_Mat.rows, OpenCV_Mat.cols, OpenCV_Mat.channels());`
> 
> and used this do vice-versa:
> 
> `Mat OpenCV_Output(Halide_Output.width(), Halide_Output.height(), CV_32F, Halide_Output.data());`
> 
> But the Halide-blurred image seems strange! I didn't change the formula [here](https://github.com/halide/CVPR2015/blob/master/blur.cpp).
> 
> ![Screenshot from 2020-05-31 13-00-01](https://user-images.githubusercontent.com/25720728/83348084-c5336100-a33e-11ea-8d64-45c94c906672.png)

check this post
https://stackoverflow.com/questions/39970361/how-to-convert-opencv-mat-to-halide-image-and-back

**Comment by benzwt** [comment]
> Thanks. I tested it and it worked faster than OpenCV. That's awesome :)
> 
> Just two more question:
> 
> 1-` Tools::save_image` cannot save float output. I don't need to save. But want to see the output. Is there a way to convert `Halide::Buffer` to `cv::Mat` (OpenCV) and vice-versa without copying the data?
> 
> 2- How can I see the Halide-generated C or C++ code?

Please refer to lesson-21
If you are using halide generator, add c_source to -e.
It will give you a xxx.halide_generated.cpp
```
./lesson_21_generate -o . -g auto_schedule_gen -f auto_schedule_true -e static_library,c_source,h,schedule target=host auto_schedule=true machine_params=32,16777216,40
```

**Comment by abadams** [comment]
I updated that stack overflow answer with the current way to do this. The issue is that opencv stores data interleaved by default. This is slow, unfortunately, but Halide::Runtime::Buffer represents this like so:

```Halide::Runtime::Buffer<float>::make_interleaved(image.data, image.cols, image.rows, image.channels());```

The catch is that the generated code will expect planar inputs by default. See tutorial lesson 16 for details of how to work with different formats.
